client/src/reports/createlinksproxymodel.cpp

```diff
@@ -0,0 +1,83 @@
+/*
+  This file is part of FatCRM, a desktop application for SugarCRM written by KDAB.
+
+  Copyright (C) 2019 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
+  Authors: Kevin Funk <kevin.funk@kdab.com>
+
+  This program is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 2 of the License, or
+  (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "createlinksproxymodel.h"
+
+#include "accountdataextractor.h"
+#include "contactdataextractor.h"
+#include "entitytreemodel.h"
+#include "itemstreemodel.h"
+#include "opportunitydataextractor.h"
+
+using namespace Akonadi;
+
+CreateLinksProxyModel::CreateLinksProxyModel(const QString &resourceBaseUrl, QObject* parent)
+    : QIdentityProxyModel(parent)
+    , mResourceBaseUrl(resourceBaseUrl)
+{
+}
+
+void CreateLinksProxyModel::setSourceModel(QAbstractItemModel* sourceModel)
+{
+    QIdentityProxyModel::setSourceModel(sourceModel);
+
+    // find the underlying ItemsTreeModel
+    auto model = sourceModel;
+    do {
+        if (auto *itemsTreeModel = qobject_cast<ItemsTreeModel *>(model)) {
+            mItemsTreeModel = itemsTreeModel;
```

Comment:
Well, mItemsTreeModel is now completely unused, so this whole method can go away. And this new solution is actually less code than the initial one :-)

Reply:
True, thanks for the review. My initial implementation indeed sucked.